MMS model agent gets panic when downloading models

/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

```
{"level":"info","ts":1605215785.7493982,"logger":"Downloader","msg":"Downloading to model dir","modelUri":"s3://torchscript","modelDir":"/mnt/models"}
{"level":"info","ts":1605215785.7495825,"logger":"modelAgent","msg":"Download model ","modelName":"cifar10","storagUri":"s3://torchscript","modelDir":"/mnt/models"}
panic: runtime error: index out of range [1] with length 1

goroutine 22 [running]:
github.com/kubeflow/kfserving/pkg/agent/storage.(*S3Provider).DownloadModel(0xc00048d9a0, 0x7ffed6e0f493, 0xb, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x1b, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/storage/s3.go:59 +0x49c
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).download(0xc000286198, 0xc000580480, 0x7, 0xc0005804a0, 0x10, 0x2, 0x3a)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:75 +0xff
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc000286198, 0xc000580480, 0x7, 0xc00045c540, 0x0, 0x0)
	github.com/kubeflow/kfserving/pkg/agent/downloader.go:51 +0x76a
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc000286180, 0xc000580480, 0x7, 0xc00045c5a0)
	github.com/kubeflow/kfserving/pkg/agent/puller.go:131 +0x2b9
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
```

**What did you expect to happen:**


**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version:
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

--- a/hack/image_patch_dev.sh
+++ b/hack/image_patch_dev.sh
@@ -40,24 +40,24 @@ data:
     {
         "image" : "${LOGGER_IMG}",
         "memoryRequest": "100Mi",
-        "memoryLimit": "1Gi",
+        "memoryLimit": "100Mi",
         "cpuRequest": "100m",
-        "cpuLimit": "1"
+        "cpuLimit": "100m"
     }
   batcher: |-
     {
         "image" : "${BATCHER_IMG}",
         "memoryRequest": "100Mi",
-        "memoryLimit": "1Gi",
+        "memoryLimit": "100Mi",
         "cpuRequest": "100m",
-        "cpuLimit": "1"
+        "cpuLimit": "100m"
     }
   agent: |-
     {
         "image" : "${AGENT_IMG}",
         "memoryRequest": "100Mi",
-        "memoryLimit": "1Gi",
+        "memoryLimit": "100Mi",
         "cpuRequest": "100m",
-        "cpuLimit": "1"
+        "cpuLimit": "100m"
     }
 EOF

--- a/pkg/agent/storage/s3.go
+++ b/pkg/agent/storage/s3.go
@@ -52,7 +52,7 @@ func (m *S3Provider) DownloadModel(modelDir string, modelName string, storageUri
 	s3Uri := strings.TrimPrefix(storageUri, string(S3))
 	tokens := strings.Split(s3Uri, "/")
 	prefix := ""
-    if len(tokens) > 1 {
+	if len(tokens) > 1 {
 		prefix = tokens[1]
 	}
 	s3ObjectDownloader := &S3ObjectDownloader{

--- a/pkg/agent/storage/s3.go
+++ b/pkg/agent/storage/s3.go
@@ -48,11 +48,11 @@ type S3ObjectDownloader struct {
 }
 
 func (m *S3Provider) DownloadModel(modelDir string, modelName string, storageUri string) error {
-	log.Info("Download model ", "modelName", modelName, "storagUri", storageUri, "modelDir", modelDir)
+	log.Info("Download model ", "modelName", modelName, "storageUri", storageUri, "modelDir", modelDir)
 	s3Uri := strings.TrimPrefix(storageUri, string(S3))
-	tokens := strings.Split(s3Uri, "/")
+	tokens := strings.SplitN(s3Uri, "/", 2)
 	prefix := ""
-	if len(tokens) > 1 {
+	if len(tokens) == 2 {
 		prefix = tokens[1]
 	}
 	s3ObjectDownloader := &S3ObjectDownloader{

--- a/pkg/agent/storage/s3.go
+++ b/pkg/agent/storage/s3.go
@@ -50,13 +50,17 @@ type S3ObjectDownloader struct {
 func (m *S3Provider) DownloadModel(modelDir string, modelName string, storageUri string) error {
 	log.Info("Download model ", "modelName", modelName, "storagUri", storageUri, "modelDir", modelDir)
 	s3Uri := strings.TrimPrefix(storageUri, string(S3))
-	path := strings.Split(s3Uri, "/")
+	tokens := strings.Split(s3Uri, "/")
+	prefix := ""
+    if len(tokens) > 1 {
+		prefix = tokens[1]
+	}
 	s3ObjectDownloader := &S3ObjectDownloader{
 		StorageUri: storageUri,
 		ModelDir:   modelDir,
 		ModelName:  modelName,
-		Bucket:     path[0],
-		Prefix:     path[1],
+		Bucket:     tokens[0],
+		Prefix:     prefix,
 		downloader: m.Downloader,
 	}
 	objects, err := s3ObjectDownloader.GetAllObjects(m.Client)

--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -33,7 +33,7 @@ type FileError error
 var NoSuccessFile FileError = fmt.Errorf("no success file can be found")
 
 func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
-	log := logf.Log.WithName("Watcher")
+	log := logf.Log.WithName("Syncer")
 	log.Info("Syncing model directory..", "modelDir", modelDir)
 	modelTracker := make(map[string]modelWrapper)
 	err := filepath.Walk(modelDir, func(path string, info os.FileInfo, err error) error {

--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -24,18 +24,17 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	"os"
 	"path/filepath"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 	"strings"
 )
-
-type Syncer struct {
-	Watcher Watcher
-}
 
 type FileError error
 
 var NoSuccessFile FileError = fmt.Errorf("no success file can be found")
 
 func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
+	log := logf.Log.WithName("Watcher")
+	log.Info("Syncing model directory..", "modelDir", modelDir)
 	modelTracker := make(map[string]modelWrapper)
 	err := filepath.Walk(modelDir, func(path string, info os.FileInfo, err error) error {
 		if info.IsDir() {

--- a/pkg/agent/watcher.go
+++ b/pkg/agent/watcher.go
@@ -50,11 +50,11 @@ func NewWatcher(configDir string, modelDir string) Watcher {
 		ModelEvents:  make(chan ModelOp, 100),
 	}
 	modelConfigFile := fmt.Sprintf("%s/%s", configDir, constants.ModelConfigFileName)
-	err  = watcher.syncModelConfig(modelConfigFile)
+	err = watcher.syncModelConfig(modelConfigFile)
 	if err != nil {
 		log.Error(err, "Failed to sync model config file")
 	}
-    return watcher
+	return watcher
 }
 
 type modelWrapper struct {

--- a/pkg/agent/watcher.go
+++ b/pkg/agent/watcher.go
@@ -18,6 +18,7 @@ package agent
 
 import (
 	"encoding/json"
+	"fmt"
 	"github.com/fsnotify/fsnotify"
 	"github.com/google/go-cmp/cmp"
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
@@ -43,16 +44,38 @@ func NewWatcher(configDir string, modelDir string) Watcher {
 	if err != nil {
 		log.Error(err, "Failed to sync model dir")
 	}
-	return Watcher{
+	watcher := Watcher{
 		configDir:    configDir,
 		modelTracker: modelTracker,
-		ModelEvents:  make(chan ModelOp),
+		ModelEvents:  make(chan ModelOp, 100),
 	}
+	modelConfigFile := fmt.Sprintf("%s/%s", configDir, constants.ModelConfigFileName)
+	err  = watcher.syncModelConfig(modelConfigFile)
+	if err != nil {
+		log.Error(err, "Failed to sync model config file")
+	}
+    return watcher
 }
 
 type modelWrapper struct {
 	Spec  *v1alpha1.ModelSpec
 	stale bool
+}
+
+func (w *Watcher) syncModelConfig(modelConfigFile string) error {
+	file, err := ioutil.ReadFile(modelConfigFile)
+	if err != nil {
+		return err
+	} else {
+		modelConfigs := make(modelconfig.ModelConfigs, 0)
+		err = json.Unmarshal(file, &modelConfigs)
+		if err != nil {
+			return err
+		} else {
+			w.parseConfig(modelConfigs)
+		}
+	}
+	return nil
 }
 
 func (w *Watcher) Start() {
@@ -82,17 +105,10 @@ func (w *Watcher) Start() {
 				if isDataDir && isCreate {
 					log.Info("Processing event", "event", event)
 					symlink, _ := filepath.EvalSymlinks(eventPath)
-					file, err := ioutil.ReadFile(filepath.Join(symlink, constants.ModelConfigFileName))
+					modelConfigFile := filepath.Join(symlink, constants.ModelConfigFileName)
+					err := w.syncModelConfig(modelConfigFile)
 					if err != nil {
-						log.Error(err, "Error in reading model config file")
-					} else {
-						modelConfigs := make(modelconfig.ModelConfigs, 0)
-						err = json.Unmarshal(file, &modelConfigs)
-						if err != nil {
-							log.Error(err, "Failed to unmarshall model config")
-						} else {
-							w.parseConfig(modelConfigs)
-						}
+						log.Error(err, "Failed to sync model config file")
 					}
 				}
 			case err, ok := <-watcher.Errors:

--- a/pkg/agent/watcher_test.go
+++ b/pkg/agent/watcher_test.go
@@ -82,9 +82,55 @@ var _ = Describe("Watcher", func() {
 		os.RemoveAll(modelDir)
 		logger.Printf("Deleted temp dir %v\n", modelDir)
 	})
-	Describe("Sync model config", func() {
-		Context("Sync new models", func() {
-			It("should download the new models", func() {
+
+	Describe("Sync models config on startup", func() {
+		Context("Getting new model events", func() {
+			It("should download and load the new models", func() {
+				defer GinkgoRecover()
+				logger.Printf("Sync model config using temp dir %v\n", modelDir)
+				watcher := NewWatcher("/tmp/configs", modelDir)
+				modelConfigs := modelconfig.ModelConfigs{
+					{
+						Name: "model1",
+						Spec: v1alpha1.ModelSpec{
+							StorageURI: "s3://models/model1",
+							Framework:  "sklearn",
+						},
+					},
+					{
+						Name: "model2",
+						Spec: v1alpha1.ModelSpec{
+							StorageURI: "s3://models/model2",
+							Framework:  "sklearn",
+						},
+					},
+				}
+				watcher.parseConfig(modelConfigs)
+				puller := Puller{
+					channelMap:  make(map[string]*ModelChannel),
+					completions: make(chan *ModelOp, 4),
+					opStats:     make(map[string]map[OpType]int),
+					Downloader: Downloader{
+						ModelDir: modelDir + "/test1",
+						Providers: map[storage.Protocol]storage.Provider{
+							storage.S3: &storage.S3Provider{
+								Client:     &mockS3Client{},
+								Downloader: &mockS3Downloder{},
+							},
+						},
+					},
+				}
+				go puller.processCommands(watcher.ModelEvents)
+				Eventually(func() int { return len(puller.channelMap) }).Should(Equal(0))
+				Eventually(func() int { return puller.opStats["model1"][Add] }).Should(Equal(1))
+				Eventually(func() int { return puller.opStats["model2"][Add] }).Should(Equal(1))
+			})
+		})
+	})
+
+	Describe("Watch model config changes", func() {
+		Context("When new models are added", func() {
+			It("Should download and load the new models", func() {
 				defer GinkgoRecover()
 				logger.Printf("Sync model config using temp dir %v\n", modelDir)
 				watcher := NewWatcher("/tmp/configs", modelDir)
@@ -126,8 +172,8 @@ var _ = Describe("Watcher", func() {
 			})
 		})
 
-		Context("Sync delete models", func() {
-			It("should download the deleted models", func() {
+		Context("When models are deleted from config", func() {
+			It("Should remove the model dir and unload the models", func() {
 				defer GinkgoRecover()
 				logger.Printf("Sync delete models using temp dir %v\n", modelDir)
 				watcher := NewWatcher("/tmp/configs", modelDir)
@@ -181,8 +227,8 @@ var _ = Describe("Watcher", func() {
 			})
 		})
 
-		Context("Sync update models", func() {
-			It("should update models", func() {
+		Context("When models uri are updated in config", func() {
+			It("Should download and reload the model", func() {
 				defer GinkgoRecover()
 				logger.Printf("Sync update models using temp dir %v\n", modelDir)
 				watcher := NewWatcher("/tmp/configs", modelDir)
@@ -218,7 +264,7 @@ var _ = Describe("Watcher", func() {
 					},
 				}
 				watcher.parseConfig(modelConfigs)
-				// remove model2
+				// update model2 storageUri
 				modelConfigs = modelconfig.ModelConfigs{
 					{
 						Name: "model1",
@@ -243,8 +289,8 @@ var _ = Describe("Watcher", func() {
 			})
 		})
 
-		Context("Model download failure", func() {
-			It("should not create success file", func() {
+		Context("When model download fails", func() {
+			It("Should not create the success file", func() {
 				defer GinkgoRecover()
 				logger.Printf("Using temp dir %v\n", modelDir)
 				var errs []s3manager.Error

--- a/pkg/apis/serving/v1beta1/predictor_triton.go
+++ b/pkg/apis/serving/v1beta1/predictor_triton.go
@@ -68,6 +68,10 @@ func (t *TritonSpec) GetContainer(metadata metav1.ObjectMeta, extensions *Compon
 	if extensions.ContainerConcurrency != nil && *extensions.ContainerConcurrency != 0 {
 		arguments = append(arguments, fmt.Sprintf("%s=%d", "--http-thread-count", *extensions.ContainerConcurrency))
 	}
+	// when storageURI is nil we enable explicit load/unload
+	if t.StorageURI == nil {
+		arguments = append(arguments, fmt.Sprintf("%s=%s", "--model-control-mode", "explicit"))
+	}
 	if t.Container.Image == "" {
 		t.Container.Image = config.Predictors.Triton.ContainerImage + ":" + *t.RuntimeVersion
 	}

